File: Python/Problem4.py
class Solution(object):
    def shortestPalindrome(self, s):
        pattern = s[::-1]  
        pattern_index = 0  
        string = s + "*" + pattern
        lps = [0] * len(string)
        for i in range(1, len(string)):
            while string[i] != string[pattern_index] and pattern_index > 0:
                pattern_index = lps[pattern_index - 1]
            if string[i] == string[pattern_index]:
                pattern_index += 1
            lps[i] = pattern_index
        palindrome = pattern[:len(pattern) - pattern_index] + s
        return palindrome

File: TypeScript/Problem4.js
var shortestPalindrome = function (s) {
    var pattern = s.split('').reverse().join(''); //The string but in reverse
    var patternIndex = 0; //An index so we can 
    var string = s + "*" + pattern; //This is the string that will help us to find the lps
    var lps = []; //Longest Propper Prefix, used in the KPM algorithm. 
    var aux; //Just an auxiliar for the index
    lps[0] = 0;
    for (var i = 1; i < string.length; i++) //This loop fill the lps array
     {
        while (string[i] != string[patternIndex] && patternIndex > 0) {
            aux = patternIndex;
            patternIndex = lps[aux - 1];
        }
        if (string[i] == string[patternIndex]) {
            patternIndex += 1;
        }
        lps[i] = patternIndex;
    }
    var palindrome = pattern.slice(0, pattern.length - patternIndex) + s; //Took the prefix and put it in front of the original string.
    return palindrome;
};
